### PR TITLE
Hyperparameter optimization

### PR DESCRIPTION
**Pull request by mateuszbaran**
I've moved the important part of https://github.com/JuliaManifolds/Manopt.jl/pull/339 here.

### CODE REVIEW

**Comment by mateuszbaran** [comment]
@kellertuer I'm not sure if we should actually attempt running the hyperparameter optimization on CI? It's fairly expensive but I will try to restrict it.

**Comment by kellertuer** [comment]
We have a few examples and/or tutorials even that are run locally and just their Markdown checked in. For example the geodesic regression (tutorial in Manopt) which requires rendering with Asymptote, which can not be done on CI.

So for any Benchmark I would do the same since it would probably run for too Lon on CI.

**Comment by mateuszbaran** [comment]
Good idea, I will try that.

**Comment by mateuszbaran** [comment]
This isn't going to work without https://github.com/JuliaManifolds/Manopt.jl/pull/361 so we should finish that first.

**Comment by mateuszbaran** [comment]
@kellertuer everything works now here :slightly_smiling_face:

**Comment by kellertuer** [comment]
Nice, I can take a closer look the next few days.

At first glance:
* running benchmark and such locally and checking that in is of course fine
* maybe a few of the code blocks could be made a bit shorter with a bit more text in between?
* The literature would be nice with something “clickable” – a DOI, an arXiv preprint, something like that.

**Comment by mateuszbaran** [comment]
I considered dividing it further but there is no obvious (to me) point at which it would make sense. Feel free to suggest which block to cut though.

File: docs/src/examples/HyperparameterOptimization.md [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+# This script requires optuna to be available through PythonCall
+# You can install it for example using
+# using CondaPkg
+# ]conda add optuna
+
+optuna = pyimport("optuna")
+
+norm_inf(M::AbstractManifold, p, X) = norm(X, Inf)
+
+struct TTsuggest_int
```
Comment (kellertuer) [inline docs/src/examples/HyperparameterOptimization.md:30]: I am not so sure what this struct and the following are, but I suppose they are merely collecting results or values? Maybe that could be mentioned?

File: docs/src/examples/HyperparameterOptimization.md [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+    categorical_suggestions::Dict{String,Int},
+)
+    tt = TracingTrial(
+        TTsuggest_int(int_suggestions),
+        TTsuggest_float(float_suggestions),
+        TTsuggest_categorical(categorical_suggestions),
+        TTreport(Float64[]),
+        TTshould_prune(),
+    )
+    od(tt)
```
Comment (kellertuer) [inline docs/src/examples/HyperparameterOptimization.md:102]: This line I do not understand – `od` is just a struct from the current definitions with quite some fields, how can it be called with one parameter?

File: docs/src/examples/HyperparameterOptimization.md [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+This example shows how to automatically select the best values of hyperparameters of optimization procedures such as retraction, vector transport, size of memory in L-BFGS or line search coefficients.
+Hyperparameter optimization relies on the [Optuna](https://optuna.org/) [AkibaSanoYanaseToshihikoTakeruKoyama:2019](@cite) Python library because it is much more advanced than similar Julia projects, offering Bayesian optimization with conditional hyperparameters and early stopping.
+
+## General definitions
+
+Here are some general definitions that you will most likely be able to directly use for your problem without any changes.
+Just remember to install `optuna`, for example using `CondaPkg` Julia library.
+
+``` julia
+using Manifolds, Manopt
```
Comment (kellertuer) [inline docs/src/examples/HyperparameterOptimization.md:16]:
```suggestion
#| output: false
using Manifolds, Manopt
```

This omits the print of the cells result, which is not so informative here.

File: docs/src/examples/HyperparameterOptimization.md [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+        CondaPkg Found dependencies: /home/mateusz/.julia/environments/v1.10/CondaPkg.toml
+        CondaPkg Dependencies already up to date
+
+The next part is your hyperparameter optimization objective. The `ObjectiveData` struct contains all relevant information about the sequence of specific problems.
+The outermost key part is the `N_range` field.
+Early stopping requires a series of progressively more complex problems.
+They will be attempted from the most simple one to the most complex one, and are specified by the values of `N` in that vector.
+
+``` julia
+mutable struct ObjectiveData{TObj,TGrad}
```
Comment (kellertuer) [inline docs/src/examples/HyperparameterOptimization.md:75]:
```suggestion
#| output: false
mutable struct ObjectiveData{TObj,TGrad}
```

File: docs/src/examples/HyperparameterOptimization.md [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+
+where $N \in [2, 16, 128, 1024, 8192, 65536]$.
+
+`obj` and `grad` are the objective and gradient, here defined as below.
+Note that gradient works in-place and variants without manifolds are also provided for easier comparison with other libraries like `Optim.jl`.
+It is easiest when problems for different values `N` can be distinguished by being defined on successively larger manifolds but the script could be modified so that it’s not necessary.
+
+`pruning_losses` and `compute_pruning_losses` are related to early pruning used in Optuna and you shouldn’t have to modify them.
+
+``` julia
```
Comment (kellertuer) [inline docs/src/examples/HyperparameterOptimization.md:123]:
````suggestion
``` julia
#| output: false
````

**Review by kellertuer — COMMENTED** [review]
Oh I added the remarks to the `md` file, sorry.

Sone code blocks are a bit long but I do not see a good idea how to shorten them.

One thing I would prefer to have is a short interpretation of the search run, what do the results mean here, which method to use? Something like `manifold = 1, retr=3` is a bit magic.

File: docs/src/examples/HyperparameterOptimization.md (hunk 1 above)
Comment (mateuszbaran) [reply docs/src/examples/HyperparameterOptimization.md:30]: They are purely technical detail for collecting reported data from a single run.

File: docs/src/examples/HyperparameterOptimization.md (hunk 2 above)
Comment (mateuszbaran) [reply docs/src/examples/HyperparameterOptimization.md:102]: It has a method written later. Anyway, `compute_pruning_losses` is another technical detail, I will move it to another cell.

**Comment by mateuszbaran** [comment]
> One thing I would prefer to have is a short interpretation of the search run, what do the results mean here, which method to use? Something like `manifold = 1, retr=3` is a bit magic.

Sure, I will add that.

File: docs/src/examples/HyperparameterOptimization.md (hunk 1 above)
Comment (kellertuer) [reply docs/src/examples/HyperparameterOptimization.md:30]: Maybe add exactly this sentence (no phrased a bit nicer maybe) somewhere then :)

File: docs/src/examples/HyperparameterOptimization.md (hunk 2 above)
Comment (kellertuer) [reply docs/src/examples/HyperparameterOptimization.md:102]: Add at least a comment what that does here, or move it to later (yeah saw it later as well then).

**Comment by mateuszbaran** [comment]
I think I've addressed all your comments but I'm not sure why the documentation preview isn't updated. Maybe we are caching too much?

**Comment by kellertuer** [comment]
I don't think we cache too much, but maybe the deployment failed? Hm, does not look like. Then I am unsure why it did not update

**Comment by kellertuer** [comment]
I See. Caching the packages speeds things up here, but seems to sometimes cause this, then.

**Review by kellertuer — APPROVED** [review]
Yes, sure.

**Pull request merged**